tests/functional/test_charm.py

```diff
@@ -0,0 +1,185 @@
+        assert results.get("return-code") == 0
+
+        results = await helper.run_action_wait(
+            apt_mirror_unit, "delete-snapshot", name=name
+        )
+        assert results.get("return-code") == 0
+
+        results = await helper.run_wait(apt_mirror_unit, check_cmd)
+        assert results.get("return-code") != 0
+
+    async def test_list_snapshots_action(self, apt_mirror_unit, helper):
+        """Test list snapshots action."""
+        results = await helper.run_action_wait(apt_mirror_unit, "list-snapshots")
+        assert results.get("return-code") == 0
+
+    async def test_synchronize_action(self, apt_mirror_unit, helper):
+        """Test synchronize action."""
+        results = await helper.run_action_wait(apt_mirror_unit, "synchronize")
+        assert results.get("return-code") == 0
+
+
+@pytest.mark.usefixtures("apt_mirror_app", "apt_mirror_unit", "helper")
+class TestCharm:
+    async def test_setup_cron_schedule(
+        self, ops_test, apt_mirror_app, apt_mirror_unit, helper
+    ):
+        """Test setup cron schedule config option."""
+        cron_schedule = "0 5 * * 1"
+        await apt_mirror_app.set_config({"cron-schedule": cron_schedule})
+        await ops_test.model.wait_for_idle(apps=["apt-mirror"])
+
+        results = await helper.run_wait(
+            apt_mirror_unit, "cat /etc/cron.d/{}".format(apt_mirror_app.name)
+        )
+        assert results.get("stdout").strip() == "{} root apt-mirror".format(
+            cron_schedule
+        )
+
+        # restore configs
+        await apt_mirror_app.reset_config(["cron-schedule"])
+
+    @pytest.mark.skip
+    async def test_remove_cron_schedule(self):
+        """Test remove cron schedule config option."""
+        # currently the charm does not support remove cron schedule for
+        # apt-mirror.
+        pass
+
+    async def test_client_access(
+        self, ops_test, apt_mirror_app, apt_mirror_unit, helper
+    ):
+        """Test remote apt server can be accessed by client."""
+        # Let's use bootstack public ppa for testing; it's very small compared
+        # to ubuntu or other os's repos.
+        url = "ppa.launchpadcontent.net/canonical-bootstack/public/ubuntu"
+        mirror_list = """\
+deb https://{0} focal main
+deb https://{0} bionic main\
+""".format(
+            url
+        )
+        await apt_mirror_app.set_config({"mirror-list": mirror_list})
+        await ops_test.model.wait_for_idle(apps=["apt-mirror"])
+
+        await apt_mirror_unit.run_action("synchronize")
+        await apt_mirror_unit.run_action("create-snapshot")
+        results = await helper.run_action_wait(apt_mirror_unit, "list-snapshots")
+        list_outputs = results.get("snapshots").strip()
+        snapshot_name = re.findall(r"snapshot-\d+", list_outputs)[0]
+
+        await apt_mirror_unit.run_action("publish-snapshot", name=snapshot_name)
+        await ops_test.model.wait_for_idle(apps=["apt-mirror"], status="active")
```

Comment:
Does this test depend on `test_publish_snapshot_action` test?

Reply:
Yes, by the time we are testing TestCharm, the TestCharmAction should all passed.

Reply:
I can do this mannually by removing some files in the machine, but there's not action to "unpublish" unfortunately. The pushlish action is simply creating a simlink of a "publish" directory to the snapshot, and if you publish again the simlink of "publish" will be overwritten.

Reply:
oh, do I understand your question correctly? The `test_client_access` *does not* need `test_publish_snapshot_aciton` to be successful.  `test_client_access` create its own snapshot and publish for client to consume testing.